File: src/rate-limiters/sliding-window.ts
import type { Redis } from 'ioredis';
import { randomUUID } from 'node:crypto';
import type { RateLimiter } from './types.js';
import type { RateLimitResult } from '../types/index.js';
import { readFileSync, existsSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));

function getLuaScript(): string {
  const p1 = resolve(__dirname, 'lua', 'sliding-window.lua');
  if (existsSync(p1)) return readFileSync(p1, 'utf-8');
  const p2 = resolve(__dirname, '..', '..', 'src', 'rate-limiters', 'lua', 'sliding-window.lua');
  if (existsSync(p2)) return readFileSync(p2, 'utf-8');
  return '';
}

const luaScript = getLuaScript();

/**
 * Sliding Window Log Rate Limiter
 *
 * Stores each request timestamp in a Redis Sorted Set (ZSET).
 */
export class SlidingWindowLimiter implements RateLimiter {
  constructor(private redis: Redis) {}

  async check(
    apiKeyId: string,
    limit: number,
    windowSecs: number,
  ): Promise<RateLimitResult> {
    const key = `ratelimit:sliding:${apiKeyId}`;

    const timeResult = await this.redis.time() as unknown as [string, string];
    const nowMs = parseInt(timeResult[0], 10) * 1000 + Math.floor(parseInt(timeResult[1], 10) / 1000);

    const requestId = `${nowMs}:${randomUUID()}`;

    let result: number[];
    if (typeof (this.redis as any).slidingWindowCheck === 'function') {
      result = await (this.redis as any).slidingWindowCheck(
        key,
        limit,
        windowSecs * 1000,
        nowMs,
        requestId,
      ) as number[];
    } else {
      result = await this.redis.eval(
        luaScript,
        1,
        key,
        limit,
        windowSecs * 1000,
        nowMs,
        requestId,
      ) as number[];
    }

    const [allowed, remaining, resetMs, currentCount] = result;

    return {
      allowed: allowed === 1,
      remaining,
      resetAt: Math.ceil(resetMs / 1000),
      currentCount,
    };
  }
}

File: src/rate-limiters/lua/sliding-window.lua
-- Sliding Window Log Rate Limiter
-- Uses a sorted set of request timestamps for precise sliding window tracking
--
-- KEYS[1] = ratelimit:sliding:{api_key_id}
-- ARGV[1] = limit
-- ARGV[2] = window_ms (window duration in milliseconds)
-- ARGV[3] = now_ms (current timestamp in milliseconds, from Redis TIME)
-- ARGV[4] = request_id (unique ID for this request, used as sorted set member)
--
-- Returns: {allowed (0/1), remaining, reset_ms, current_count}

local key = KEYS[1]
local limit = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local request_id = ARGV[4]

-- Remove all entries older than the sliding window
redis.call('ZREMRANGEBYSCORE', key, 0, now - window)

-- Count current entries in the window
local current = redis.call('ZCARD', key)
local allowed = 0

if current < limit then
  -- Add this request's timestamp as score, request_id as member
  redis.call('ZADD', key, now, request_id)
  current = current + 1
  allowed = 1
end

-- Set expiry on the key for cleanup (window duration)
redis.call('PEXPIRE', key, window)

local remaining = limit - current
if remaining < 0 then
  remaining = 0
end

-- Reset time = oldest entry timestamp + window duration
local oldest = redis.call('ZRANGE', key, 0, 0, 'WITHSCORES')
local reset = now + window
if oldest and oldest[2] then
  reset = tonumber(oldest[2]) + window
end

return { allowed, remaining, reset, current }
